### PR TITLE
Max7219 in Dot Matrix configuration

### PR DESCRIPTION
**Pull request by rspaargaren**
## Description:

**Related issue (if applicable):** fixes 
https://github.com/esphome/feature-requests/issues/14#

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#585

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by glmnet** [comment]
a link in index.rst is missing

**Comment by rspaargaren** [comment]
index.rst has been added to the documentation

File: esphome/components/max7219digit/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+https://user-images.githubusercontent.com/10338724/81001535-2c491d00-8e48-11ea-86f2-2b0bb60eacb0.jpeg
```
Comment (glmnet) [inline esphome/components/max7219digit/README.md:1]: remove this file, we don't add readmes for components, they all have their docs in their place, if some info here is relevant it should be included in the docs.

File: esphome/components/max7219digit/display.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+CONFIG_SCHEMA = display.BASIC_DISPLAY_SCHEMA.extend({
+    cv.GenerateID(): cv.declare_id(MAX7219Component),
+    cv.Optional(CONF_NUM_CHIPS, default=4): cv.int_range(min=1, max=255),
+    cv.Optional(CONF_INTENSITY, default=15): cv.int_range(min=0, max=15),
+    cv.Optional(CONF_ROTATE_CHIP, default=0): cv.int_range(min=0, max=4),
+    cv.Optional(CONF_SCROLL_MODE, default=0): cv.int_range(min=0, max=1),
+    cv.Optional(CONF_SCROLL_ENABLE, default=True): cv.boolean,
+    cv.Optional(CONF_SCROLL_SPEED, default=250): cv.int_range(min=0),
+    cv.Optional(CONF_SCROLL_DELAY, default=1000): cv.int_range(min=0),
+    cv.Optional(CONF_SCROLL_DWELL, default=1000): cv.int_range(min=0),
```
Comment (glmnet) [inline esphome/components/max7219digit/display.py:29]: these should be time type, see https://deploy-preview-585--esphome.netlify.app/guides/configuration-types.html#config-time and examples https://github.com/esphome/esphome/blob/fed37b48a107dbf01b4cf53b3906b8a3a5e64c46/esphome/components/time_based/cover.py#L18

File: esphome/components/max7219digit/display.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+MAX7219Component = max7219_ns.class_('MAX7219Component', cg.PollingComponent, spi.SPIDevice,
+                                     display.DisplayBuffer)
+MAX7219ComponentRef = MAX7219Component.operator('ref')
+
+CONFIG_SCHEMA = display.BASIC_DISPLAY_SCHEMA.extend({
+    cv.GenerateID(): cv.declare_id(MAX7219Component),
+    cv.Optional(CONF_NUM_CHIPS, default=4): cv.int_range(min=1, max=255),
+    cv.Optional(CONF_INTENSITY, default=15): cv.int_range(min=0, max=15),
+    cv.Optional(CONF_ROTATE_CHIP, default=0): cv.int_range(min=0, max=4),
+    cv.Optional(CONF_SCROLL_MODE, default=0): cv.int_range(min=0, max=1),
```
Comment (glmnet) [inline esphome/components/max7219digit/display.py:25]: these should be an enum

File: tests/max7219-2.yaml [hunk 4]
```diff
@@ -0,0 +1,95 @@
+esphome:
```
Comment (glmnet) [inline tests/max7219-2.yaml:1]: remove unrelated files, only 1 test file is required for this integration, the goal is to have the cpp code being compiled, not much testing is possible with different yamls

File: tests/nextion/partitions.csv [hunk 5]
```diff
@@ -0,0 +1,6 @@
+nvs,      data, nvs,     0x009000, 0x005000,
```
Comment (glmnet) [inline tests/nextion/partitions.csv:1]: remove this

File: esphome/components/max7219digit/max7219digit.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+ protected:
+  void send_byte_(uint8_t a_register, uint8_t data);
+  void send_to_all_(uint8_t a_register, uint8_t data);
+
+  uint8_t intensity_{15};  /// Intensity of the display from 0 to 15 (most)
+  uint8_t num_chips_{1};
+  bool scroll_{false};
+  bool update_{false};
+  uint16_t scroll_speed_ = 250;
```
Comment (glmnet) [inline esphome/components/max7219digit/max7219digit.h:91]: remove all default values, defaults are in the py file and all setters are called. Cpp code is not supported without py generation

File: esphome/components/max7219digit/max7219digit.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+}
+void MAX7219Component::set_writer(max7219_writer_t &&writer) { this->writer_local_ = writer; }
+void MAX7219Component::set_intensity(uint8_t intensity) { this->intensity_ = intensity; }
+void MAX7219Component::set_num_chips(uint8_t num_chips) { this->num_chips_ = num_chips; }
+void MAX7219Component::set_chip_orientation(uint8_t rotate) { this->orientation_ = rotate; }
+void MAX7219Component::set_scroll_speed(uint16_t speed) { this->scroll_speed_ = speed; }
+void MAX7219Component::set_scroll_dwell(uint16_t dwell) { this->scroll_dwell_ = dwell; }
+void MAX7219Component::set_scroll_delay(uint16_t delay) { this->scroll_delay_ = delay; }
+void MAX7219Component::set_scroll(bool on_off) { this->scroll_ = on_off; }
+void MAX7219Component::set_scroll_mode(uint8_t mode) { this->scroll_mode_ = mode; }
```
Comment (glmnet) [inline esphome/components/max7219digit/max7219digit.cpp:292]: you can implement all this in the h if you want

**Review by glmnet — CHANGES_REQUESTED** [review]
Looking good!

File: esphome/components/max7219digit/README.md (hunk 1 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/README.md]: Readme file has been removed. This was not up to date anyhow...

File: esphome/components/max7219digit/display.py (hunk 2 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/display.py]: Type has been adjusted to time_period and default values have been updated to new format

File: tests/max7219-2.yaml (hunk 4 above)
Comment (rspaargaren) [reply tests/max7219-2.yaml]: Only one test file MAX7219.yaml has been left over in test dir

File: tests/nextion/partitions.csv (hunk 5 above)
Comment (rspaargaren) [reply tests/nextion/partitions.csv:1]: REMOVED was left over from previous edit

File: esphome/components/max7219digit/max7219digit.h (hunk 6 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.h]: Default values which are set in Python config have been removed from the file.

File: esphome/components/max7219digit/max7219digit.cpp (hunk 7 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp]: These have  been moved to the h file.

File: esphome/components/max7219digit/display.py (hunk 3 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/display.py]: It was a bit of searching on how to do but enum types have been added and the doc files have been updated.

File: esphome/components/max7219digit/display.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+max7219_ns = cg.esphome_ns.namespace('max7219digit')
+MAX7219Component = max7219_ns.class_('MAX7219Component', cg.PollingComponent, spi.SPIDevice,
+                                     display.DisplayBuffer)
+MAX7219ComponentRef = MAX7219Component.operator('ref')
+
+CONFIG_SCHEMA = display.BASIC_DISPLAY_SCHEMA.extend({
+    cv.GenerateID(): cv.declare_id(MAX7219Component),
+    cv.Optional(CONF_NUM_CHIPS, default=4): cv.int_range(min=1, max=255),
+    cv.Optional(CONF_INTENSITY, default=15): cv.int_range(min=0, max=15),
+    cv.Optional(CONF_ROTATE_CHIP, default='CONT'): cv.enum(CHIP_MODES, upper=True),
```
Comment (glmnet) [inline esphome/components/max7219digit/display.py:36]: default value is not a CHIP_MODES value, is causing to fail when default value is tried to be used.

File: esphome/components/max7219digit/display.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+CONF_ROTATE_CHIP = 'rotate_chip'
+CONF_SCROLL_SPEED = 'scroll_speed'
+CONF_SCROLL_DWELL = 'scroll_dwell'
+CONF_SCROLL_DELAY = 'scroll_delay'
+CONF_SCROLL_ENABLE = 'scroll_enable'
+CONF_SCROLL_MODE = 'scroll_mode'
+
+SCROLL_MODES = {
+    'CONT': 0,
```
Comment (glmnet) [inline esphome/components/max7219digit/display.py:16]: change to CONTINUOUS

File: tests/test1.yaml.save [hunk 10]
```diff
@@ -0,0 +1,43 @@
+esphome:
```
Comment (glmnet) [inline tests/test1.yaml.save:1]: what's this file? save?

File: tests/max7219.yaml [hunk 11]
```diff
@@ -0,0 +1,86 @@
+esphome:
```
Comment (glmnet) [inline tests/max7219.yaml:1]: just an entry on any of the test123.yaml files is ok, we don't use this, all we validate is the cpp code is compiled ok

File: esphome/components/max7219digit/max7219digit.cpp [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+
+void MAX7219Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "MAX7219DIGIT:");
+  ESP_LOGCONFIG(TAG, "  Number of Chips: %u", this->num_chips_);
+  ESP_LOGCONFIG(TAG, "  Intensity: %u", this->intensity_);
+  ESP_LOGCONFIG(TAG, "  Scroll Mode: %u", this->scroll_mode_);
+  ESP_LOGCONFIG(TAG, "  Scroll Speed: %u", this->scroll_speed_);
+  ESP_LOGCONFIG(TAG, "  Scroll Dwell: %u", this->scroll_dwell_);
+  ESP_LOGCONFIG(TAG, "  Scroll Delay: %u", this->scroll_delay_);
+  // ESP_LOGCONFIG(TAG, "  Offset: %u", this->offset_chips_);
```
Comment (glmnet) [inline esphome/components/max7219digit/max7219digit.cpp:53]: we is commented? is it going to be added in the future?

**Review by glmnet — CHANGES_REQUESTED** [review]

File: tests/test1.yaml.save (hunk 10 above)
Comment (rspaargaren) [reply tests/test1.yaml.save:1]: no use is removed

File: esphome/components/max7219digit/max7219digit.cpp (hunk 12 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp:53]: No this was left over from past, removed

File: esphome/components/max7219digit/display.py (hunk 8 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/display.py:36]: changed to 0

File: esphome/components/max7219digit/display.py (hunk 9 above)
Comment (rspaargaren) [reply esphome/components/max7219digit/display.py:16]: done

File: tests/max7219.yaml (hunk 11 above)
Comment (rspaargaren) [reply tests/max7219.yaml:1]: entry has been added to test3.yaml

**Comment by glmnet** [comment]
@igg I'm checking this component and it works with CS_PIN set, this code does not seem to do anything with the cs_pin explicitely, so I just tested it with the cs_pin to gnd and to +vcc and none of them works, it only works if it is connected.
This PR is ok because is requesting the cs pin to be connected, but I'm wondering if we didn't something wrong by changing other components too.

**Review by glmnet — CHANGES_REQUESTED** [review]
Try all your methods which are in your docs, the one of the image does not work, some lines are rendered off buffer. I'm not going to add review on docs yet, but there are changes needed too.

File: esphome/components/max7219digit/max7219digit.cpp [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+    this->max_displaybuffer_.resize(x + 1, this->bckgrnd_);
+  }
+
+  if (y >= this->get_height_internal() || y < 0)  // If pixel is outside display then dont draw
+    return;
+
+  uint16_t pos = x;    // X is starting at 0 top left
+  uint8_t subpos = y;  // Y is starting at 0 top left
+
+  if (this->invert_) {
```
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp:135]: I don't fully understand your remark. Why should I use the color info for drawing a pixel? In the display buffer the Draw pixel routine in only called if a pixel has to be drawn. And the max7219 draws a pixel independently of the color as it has no gradient... The routine works...

File: esphome/components/max7219digit/max7219digit.cpp [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+  }
+}
+
+void MAX7219Component::display() {
+  byte pixels[8];
+  for (uint8_t i = 0; i < this->num_chips_; i++) {  // Run this loop for every MAX CHIP (GRID OF 64 leds)
+    for (uint8_t j = 0; j < 8; j++) {  // Run this routine for the rows of every chip 8x row 0 top to 7 bottom
+      pixels[j] = this->max_displaybuffer_[i * 8 + j];  // Fill the pixel parameter with diplay data
+    }
+    this->send64pixels(i, pixels);  // Send the data to the chip
```
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp:111]: Some of the not always logic spaces etc is to keep travis happy...
I can remove some commend, but to my opinion they are usefull to understand what is happening for future editing...?
I will check in the invert routine. This did not have to much focus

**Comment by glmnet** [comment]
Try the image example.  The smiley

**Comment by igg** [comment]
> This PR is ok because is requesting the cs pin to be connected, but I'm wondering if we didn't something wrong by changing other components too.

If there is no writing to the CS pin, then how can it matter if it is connected to ground or connected to a pin which is set to output 0. It's probably worth trying to figure out why it makes a difference.
Some pins on the ESP undergo a transition from reset/boot to running code. Does the max7219 datasheet say anything about required transitions on the CS pin like the as3935?

**Comment by rspaargaren** [comment]
> Try the image example. The smiley

Yes, understood. Indeed removed the color during coding not realizing that a image was actually using it... repaired it in the current code.

Invert work now simple but the code needs some clean up...

File: esphome/components/max7219digit/max7219digit.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp]: See remark below. Repaired did not use those function for a while.

File: esphome/components/max7219digit/max7219digit.cpp [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+}
+void MAX7219Component::update() {
+  this->update_ = true;
+  this->max_displaybuffer_.clear();
+  this->max_displaybuffer_.resize(this->num_chips_ * 8, this->bckgrnd_);
+  if (this->writer_local_.has_value())  // insert Labda function if available
+    (*this->writer_local_)(*this);
+}
+
+void MAX7219Component::invert_on_off(bool on_off) {
```
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp:160]: Will clean up the invert code. As routine is now simple and does not require these codes

**Comment by rspaargaren** [comment]
> Try all your methods which are in your docs, the one of the image does not work, some lines are rendered off buffer. I'm not going to add review on docs yet, but there are changes needed too.

Can you give me a hint what needs to be changed on the docs. So I can work on that in parallel...

**Comment by glmnet** [comment]
I’ve already made some changes. 
But the examples are wrong. 
Let me comment there ...

File: esphome/components/max7219digit/max7219digit.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp]: I have cleaned up a bit but most of the formatting is dictated by travis to make it pass

File: esphome/components/max7219digit/max7219digit.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp]: invert code is now very simple and buffer is just buffer. Thanks for point me on that.

File: esphome/components/max7219digit/max7219digit.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (rspaargaren) [reply esphome/components/max7219digit/max7219digit.cpp]: The color bit 1 or 0 is now used for writing the buffer as it is supposed to do...

**Review by glmnet — CHANGES_REQUESTED** [review]
Please also check my comments on the docs, PR.
It's almost there!!!

File: .gitignore [hunk 16]
```diff
@@ -118,3 +118,10 @@ config/
 tests/build/
 tests/.esphome/
 /.temp-clang-tidy.cpp
+tests/maxsmile.yaml
+tests/pixelmix_bold.ttf
+tests/pixelmix.ttf
+tests/pixelmix_ss.ttf
+tests/smile.png
+tests/nextion/partitions.csv
+.gitignore
```
[the remark this answers is not in the document]
Comment (rspaargaren) [reply .gitignore]: I have reverted the changes to .gitignore

**Comment by rspaargaren** [comment]
> Please also check my comments on the docs, PR.
> It's almost there!!!

Yes hopefully final round ;-)

**Pull request merged**